File: data/mi1/rooms/scummbar.lua
require("template/room1")

roomDefinition = {
	width = 640,
	height = 144,
	startTable = {
		village1 = { pos = items2["scummbar.door_out"].hotspot.walk_to, dir = "east"},
		--mancomb = { pos = items2["scummbar.mancomb"].hotspot.walk_to, dir = "north"},
		--estevan = { pos = items2["scummbar.estevan"].hotspot.walk_to, dir = "south"},
		--kitchen = { pos = items2["scummbar.door_kitchen"].hotspot.walk_to, dir="west" }

	},
	defaultroom = "village1",
	depth = { type="linear_y", values= {0, 1, 144, 0} },
	--depth = { type="linear_y", values= {0, 1, 144, 0} },
	--scale = { type="linear_y", values= {0, 0.8, 144, 0.2}}
}

room = generateBasicRoom (roomDefinition)

-- room:add_asset(sprites.door_scummbar_village)
-- room:add_asset(sprites.door_scummbar_kitchen)
-- room:add_asset(sprites.mancomb)
-- room:add_asset(sprites.estevan)
-- room:add_asset(sprites.loompirate)
-- room:add_asset(sprites.fireplace)
-- room:add_asset(sprites.cook)
-- room:add_asset(sprites.ilp1)
-- room:add_asset(sprites.ilp2)
-- room:add_asset(sprites.ilp3)
variables[items2["scummbar.door_kitchen"].variable] = 0


room:add( {
	{ pos = {0, 0,-3}, components = { { type="gfx", image="gfx/scummbar_1.png" }}},
	factory.walkarea { shape = { type = "poly", outline = {32,16,70,24,128,19,251,
		18,311,10,321,10,345,32,467,41,492,50,514,40,565,40,580,35,629,6,626,0,256,0,200,16,149,0,90,0,85,10},
		holes = {
			{374,6,505,6,505,28,374,28}
		}
	}},
	factory.objc { id="scummbar.door_out" },
	factory.objc { id="scummbar.door_kitchen" },


	-- { pos = {157, 0, 0.99}, components = { { type="gfx", image="gfx/scummbar_3.png" }}},
	-- { pos = {20, 0, 0.99}, components = { { type="gfx", image="gfx/scummbar_4.png" }}},
	-- { pos = {374, 20, 0.95}, components = { { type="gfx", image="gfx/scummbar_2.png" }}},
	-- factory.object.create { object = "scummbar.door_out" },
	-- factory.object.create { object = "scummbar.mancomb" },
	-- factory.object.create { object = "scummbar.estevan" },
	-- factory.object.create { object = "scummbar.loompirate"},
	-- factory.object.create { object = "scummbar.fireplace"},
	-- factory.object.create { object = "scummbar.ilp1"},
	-- factory.object.create { object = "scummbar.ilp2"},
	-- factory.object.create { object = "scummbar.ilp3"},
	-- factory.object.create { object = "scummbar.important_looking_pirates"},
	-- factory.walkarea.create {
 --   		shape = { 
	--     	type = "poly", 
	--         outline = {32,16,70,24,128,19,251,18,311,10,321,10,345,32,467,41,492,50,514,40,565,40,580,35,629,6,626,0,256,0,200,16,149,0,90,0,85,10},
	--         holes = {
	--         	{374,6,505,6,505,28,374,28}
	--         }
	--     }
	-- }

})


function run_background_script(actor, anim) 
	local mancomb_script = script:new()
	mancomb_script.actions = {
		action.random_delay { id=1, min=1, max=4 },
		action.animate_once { id=2, actor = actor, anim = anim },
		action.animate { id=3, actor = actor, anim = "idle" },
	}
	mancomb_script.loop = 1
	monkey.play(mancomb_script)
end

function run_background_script_2(actor, anim_transition, anim2) 
	local mancomb_script = script:new()
	mancomb_script.actions = {
		action.random_delay { id=1, min=1, max=4 },
		action.animate_once { id=2, actor = actor, anim = anim_transition },
		action.animate { id=3, actor = actor, anim = anim2, },
		action.random_delay { id=4, min=1, max=4 },
		action.animate_once { id=5, actor = actor, anim = anim_transition },		
		action.animate { id=6, actor = actor, anim = "idle", },
		
	}
	mancomb_script.loop = 1
	monkey.play(mancomb_script)
end

local cook = function() 
	--local s = script:new("cook")
	local pos = items2["scummbar.door_kitchen"].hotspot.walk_to
	-- coming from village
	local actions = {
		{ type = action.delay, args = {sec=10}},
		{ type = action.open_door, args = {door="scummbar.door_kitchen"}},
		{ type = action.set_variable, args = {var = "cook_in_kitchen", value = false }},
		{ type = action.create_object, args = {name="scummbar.cook", pos = {pos[1], pos[2], 0} }}


	}

	local s = ms2(actions)
	monkey.play(s)

	-- if (variables._previousroom == "kitchen") then
	-- 	local mancombPos = items["scummbar.mancomb"].walk_to
	-- 	variables["cook_in_kitchen"] = false
	-- 	s.actions = {
	-- 		action.create_object { id = 1, name="scummbar.cook", pos = {mancombPos[1], mancombPos[2], 0}, anim ="idle_back", face = "north", applydepth = true},
	-- 		action.delay { id=2, sec = 5 },
	-- 		action.walkto { id=3, actor ="scummbar.cook", obj = items["scummbar.door_kitchen"]},
	-- 		action.remove_object { id = 4, name = "scummbar.cook" },
	-- 		action.close_door { id = 5, door = "scummbar.door_kitchen"},
	-- 		action.set_variable { id = 6, var = "cook_in_kitchen", value = true },
	-- 	}
	-- end
	-- local s1 = script:new()
	-- s1.actions = {
	-- 	action.delay { id=1, sec = 10 },
	-- 	action.open_door { id = 2, door = "scummbar.door_kitchen"},
	-- 	action.set_variable { id = 3, var = "cook_in_kitchen", value = false },
	-- 	action.create_object { id = 4, name="scummbar.cook", pos = {pos[1], pos[2], 0}, anim ="idle_right", face = "west", applydepth = true},
	-- 	action.walkto { id=5, actor ="scummbar.cook", obj = items["scummbar.mancomb"]},
	-- 	action.turn { id = 6, actor="scummbar.cook", dir="north"},
	-- 	action.delay { id=7, sec = 5 },
	-- 	action.walkto { id=8, actor ="scummbar.cook", obj = items["scummbar.door_kitchen"]},
	-- 	action.remove_object { id = 9, name = "scummbar.cook" },
	-- 	action.close_door { id = 10, door = "scummbar.door_kitchen"},
	-- 	action.set_variable { id = 11, var = "cook_in_kitchen", value = true },
	-- }
	-- s1.loop = 1
	-- s:push{script=s1}
	-- monkey.play(s)

end


function room.afterstartup() 
	for k, v in ipairs(room.initstuff) do
		v()
	end
	cook()
	--run_background_script ("scummbar.mancomb", "drink")
	--run_background_script ("scummbar.estevan", "drink")
	--run_background_script_2 ("scummbar.loompirate", "move", "idle2")
	--scripts.cook()
end

File: data/mi1_py/rooms/scummbar.py
import lib_py.engine as engine
import lib_py.scumm.room as room
import lib_py.scumm.actions as act
import lib_py.actions as actions
import lib_py.scumm.entity as se
import lib_py.shape as sh
from lib_py.script import Script
from lib_py.scumm.scumm import State
import lib_py.scumm.scripts as scr
import mi1_py.variables as var
import example    



def animate():
    run_background_script('scummbar.mancomb', 1, 4, 0, 0)
    run_background_script('scummbar.pirate1', 1, 4, 1, 4)
    run_background_script('scummbar.pirate2', 1, 4, 1, 4)
    run_background_script("scummbar.pirate3", 1, 4, 0.5, 0.5)
    run_background_script("scummbar.pirate4", 1, 4, 0, 0)
    run_background_script("scummbar.pirate5", 1, 4, 1, 2)
    run_background_script("scummbar.pirate7", 1, 4, 1, 2)
    run_background_script("scummbar.pirate8", 1, 4, 0, 0)
    run_background_script("scummbar.pirate9", 1, 4, 0, 0)
    run_background_script("scummbar.estevan", 1, 4, 0, 0)
    run_background_script("scummbar.loom", 1, 4, 0, 0)

def run_background_script(actor, a, b, c, d):
    s = Script(loop = 0)
    s.addAction (actions.DelayRandom(min=a, max=b))
    s.addAction (actions.Animate(tag=actor, anim='idle_2', fwd = True, sync = True))
    s.addAction (actions.DelayRandom(min=c, max=d))
    s.addAction (actions.Animate(tag=actor, anim='idle_2', fwd = False, sync = True))
    s.addAction (actions.Animate(tag=actor, anim='idle_1', fwd = True, sync = True))
    example.play(s)

def makeCook(pos, dir):
    return se.Character (
        model = 'cook', 
        speed = 100, 
        dir = dir, 
        state='idle', 
        text_color = [85, 85, 255, 255], 
        text_offset = [0,60], 
        tag = 'cook',
        pos = pos)


def cook():
    def setCookInKitchen():
        var.cook_in_kitchen = True

	# if coming from kitchen, cook should be added to the scene!
    cook_pos = [125, 17]
    if engine.previous_room == 'kitchen':
        var.cook_in_kitchen = False
        s = Script()
        # make the cook
        cook = makeCook (cook_pos, 'n')
        s.addAction (actions.AddEntity (cook, 'walkarea'))
        s.addAction (actions.Delay (sec = 5))
        s.addAction (act.Walk (tag = 'cook', pos = var.scummbar_kitchen_door_pos))
        s.addAction (actions.RemoveEntity(tag = 'cook'))
        s.addAction (actions.CallFunc(f = setCookInKitchen))
        example.play(s)
    else:
        var.cook_in_kitchen = True
        cook = makeCook (var.scummbar_kitchen_door_pos, 'w')
        s = Script()
        s.addAction (action = actions.Delay (sec=10), id = 0)
        s.addAction (actions.RunScript (s = scr.openDoor(doorId='scummbar.door.kitchen', var='scummbar_kitchen')() ))
        s.addAction (actions.AddEntity (cook, 'walkarea'))
        s.addAction (act.Walk (tag = 'cook', pos = cook_pos))
        s.addAction (act.Turn (tag = 'cook', dir= 'n'))
        s.addAction (actions.Delay (sec=5))
        s.addAction (act.Walk (tag = 'cook', pos = var.scummbar_kitchen_door_pos))
        s.addAction (actions.RemoveEntity(tag = 'cook'))
        s.addAction (actions.RunScript (s = scr.closeDoor(doorId='scummbar.door.kitchen', var='scummbar_kitchen')() ))
        s.loop = 0
        example.play(s)
#     local a2 = {
# 		{ ref = 1, type = action.delay, args = {sec = 10} },
# 	 	mi.script.open_door { door='scummbar.door.kitchen', value=true },
# 	 	{ type = action.set_variable, args = { var = 'cook_in_kitchen', value = false }},
# 		{ type = action.create_object, args = { 
# 			factory = scumm.factory.object, 
# 			args = { id="scummbar.cook", params = { pos = mi.rooms.scummbar.door_kitchen, dir='w' } }, 
# 			parent = "scummbar.walkarea", 				
# 		}},
# 		{ type = scumm.action.walkto, args = { tag = 'scummbar.cook', obj = "scummbar.mancomb" }}, --obj = items["scummbar.mancomb"]},		
# 		{ type = scumm.action.turn, args = { tag = 'scummbar.cook', dir='north'}},
# 	 	{ type = action.delay, args = {sec = 5 } },
# 	 	{ type = scumm.action.walkto, args = { tag ="scummbar.cook", obj = 'scummbar.door.kitchen'}},
# 	 	{ type = action.remove_object, args = { tag = 'scummbar.cook'}},
# 	 	mi.script.open_door { door='scummbar.door.kitchen', value=false },
# 		{ type = action.set_variable, args = {var = 'cook_in_kitchen', value = true }},
# 	 }


# 	 	a1 = {
# 	 		{ type = action.create_object, args = { 
# 	 			factory = scumm.factory.object, 
# 	 			args = { id = 'scummbar.cook', params = { pos = {mancombPos[1], mancombPos[2], 0}, dir='n' }},
# 	 			parent = 'scummbar.walkarea'
# 	 		}},
# 	 		{ type = action.delay, args = {sec = 5 }},
# 	 		{ type = scumm.action.walkto, args = { tag ='scummbar.cook', obj = 'scummbar.door.kitchen'}},
# 	 		{ type = action.remove_object, args = { tag ="scummbar.cook" }},
# 	 		mi.script.open_door { door='scummbar.door.kitchen', value = false },
# 	 		{ type = action.set_variable, args = {var = 'cook_in_kitchen', value = true }}
# 	 	}
# 	else
# 		variables.cook_in_kitchen = true
# 	end
# 	-- 	variables[items["scummbar.door_kitchen"].variable] = 0
# 	-- end
#     local a2 = {
# 		{ ref = 1, type = action.delay, args = {sec = 10} },
# 	 	mi.script.open_door { door='scummbar.door.kitchen', value=true },
# 	 	{ type = action.set_variable, args = { var = 'cook_in_kitchen', value = false }},
# 		{ type = action.create_object, args = { 
# 			factory = scumm.factory.object, 
# 			args = { id="scummbar.cook", params = { pos = mi.rooms.scummbar.door_kitchen, dir='w' } }, 
# 			parent = "scummbar.walkarea", 				
# 		}},
# 		{ type = scumm.action.walkto, args = { tag = 'scummbar.cook', obj = "scummbar.mancomb" }}, --obj = items["scummbar.mancomb"]},		
# 		{ type = scumm.action.turn, args = { tag = 'scummbar.cook', dir='north'}},
# 	 	{ type = action.delay, args = {sec = 5 } },
# 	 	{ type = scumm.action.walkto, args = { tag ="scummbar.cook", obj = 'scummbar.door.kitchen'}},
# 	 	{ type = action.remove_object, args = { tag = 'scummbar.cook'}},
# 	 	mi.script.open_door { door='scummbar.door.kitchen', value=false },
# 		{ type = action.set_variable, args = {var = 'cook_in_kitchen', value = true }},
# 	 }
	
# 	local actions = {}
# 	table.insert(actions, a1)
# 	table.insert(actions, a2)
# 	--local s = ms2(actions, 1)

# 	local s = script.make(actions, 1)
# 	s.name = "_cook"
# 	monkey.play(s)
# end

def builder():
    r =  room.RoomUI(id='scummbar', width = 640, height = 144)
    r.add (se.BackgroundItem(image='gfx/scummbar_1.png', pos = [0, 0, -3]), 'main')
    r.add (se.BackgroundItem(image='gfx/scummbar_2.png', pos = [374, 20, 0.95]), 'main')
    r.add (se.BackgroundItem(image='gfx/scummbar_3.png', pos = [157, 0, 0.99]), 'main')
    r.add (se.BackgroundItem(image='gfx/scummbar_4.png', pos = [20, 0, 0.99]), 'main')

    r.add (se.WalkArea (tag='walkarea', shape = sh.Polygon(
        outline = [32, 16, 70, 24, 128, 19, 251, 18, 311, 10, 321, 10, 345, 32, 467, 41,
 			492, 50, 514, 40, 565, 40, 580, 35, 629, 6, 626, 0, 256, 0, 200, 16, 149, 0, 90, 0, 85, 10],
        holes = [ [374, 28, 505, 28, 505, 6, 374, 6]]),
        depth = sh.LinY(y0=0,z0=1,y1=144,z1=0)), 'main')
    
    # doors
    r.add (se.Sprite (item='scummbar.door.out', anim = var.doors.village_scummbar, model='door_scummbar_village', pos = [32, 24, 0]), 'main')	
    r.add (se.Sprite (item='scummbar.door.kitchen', anim = var.doors.scummbar_kitchen, model='door_scummbar_kitchen', pos = [591, 9, 0]), 'main')	
    
    # other
    r.add (se.Sprite (model = 'scummbar.fireplace', item = 'scummbar.fireplace', pos = [509, 44,1]), 'walkarea')


    # pirates
    r.add (se.Sprite (item = 'scummbar.mancomb', model = 'scummbar.mancomb', pos = [89, 24, -1]), 'main')
    r.add (se.Sprite (item = 'scummbar.estevan', model = 'scummbar.estevan', pos = [164, 21, 1]), 'main')
    r.add (se.Sprite (item = 'scummbar.loom', model = 'scummbar.loom', pos = [260, 17, 1]), 'walkarea')
    r.add (se.Sprite (item = 'scummbar.pirate1', model = 'scummbar.pirate1', pos = [138, 38, -1]), 'main')
    r.add (se.Sprite (model = 'scummbar.pirate2', pos = [124,20,-1], tag = 'scummbar.pirate2'), 'main')
    r.add (se.Sprite (model = 'scummbar.pirate3', pos = [30,0,-1], tag = 'scummbar.pirate3'), 'walkarea')
    r.add (se.Sprite (item = 'scummbar.pirate4', model = 'scummbar.pirate4', pos = [0, 0, -1]), 'walkarea')
    r.add (se.Sprite (model = 'scummbar.pirate5', pos = [200,18,1], tag = 'scummbar.pirate5'), 'main')
    r.add (se.Sprite (model = 'scummbar.pirate6', pos = [160,0,1], tag = 'scummbar.pirate6'), 'main')
    r.add (se.Sprite (model = 'scummbar.pirate7', pos = [255,38,1], tag = 'scummbar.pirate7'), 'walkarea')
    r.add (se.Sprite (model = 'scummbar.pirate8', pos = [215,44,1], tag = 'scummbar.pirate8'), 'walkarea')
    r.add (se.Sprite (model = 'scummbar.pirate9', pos = [278,40,1], tag = 'scummbar.pirate9'), 'walkarea')
    r.add (se.Character (model='ilp1', speed = 100, dir = 'e', 
        state='idle', text_color=[85, 85, 255, 255], text_offset=[0,60], pos = [376, 11, 0.95]), 'walkarea')
    r.add (se.Character (model='ilp2', speed = 100, dir = 'e', 
        state='idle', text_color=[255, 255, 85, 255], text_offset=[0,60], pos = [413,11,0.95]), 'walkarea')
    r.add (se.Character (model='ilp3', speed = 100, dir = 'e', 
        state='idle', text_color=[255, 85, 255, 255], text_offset=[0,60], pos = [444,18,0.95]), 'walkarea')
    r.add (se.Sprite (item = 'scummbar.important_looking_pirates', pos=[370,30,0]), 'main')

    r.init.append (animate)
    r.init.append (cook)
    r.addDynamicItems()
    return r

engine.addRoom (id = 'scummbar', f=builder)
